src/Voron/Data/Containers/Container.cs

```diff
@@ -263,30 +263,45 @@ private static void MoveToNextPage(LowLevelTransaction llt, long containerId, re
             container.Header.OnFreeList = false; // we take it out now..., we'll add to the free list when we delete from it
             var rootPage = llt.ModifyPage(containerId);
             var rootContainer = new Container(rootPage);
-            
 
-            while (true)
-            {
+            // We wont work as hard if we know that the entry is too big.
+            bool isBigEntry = size >= (Constants.Storage.PageSize / 6);
+            int tries = isBigEntry ? 4 : 128;
+
+            // PERF: Even if this condition never happens, we need the code to ensure that we have a bounded time to find a free page.
+            // This is the case where at some point we need to just give up or end up wasting more time to find a page than the time
+            // we will use to create and store in disk a new one.
+            int i = 0;
+            for (; i < tries; i++)  
+            {                
                 var freeListStateSpan = rootContainer.GetItem(ContainerPageHeader.FreeListOffset);
                 var freeList = new Set(llt, FreePagesSetName, MemoryMarshal.AsRef<SetState>(freeListStateSpan));
                 var it = freeList.Iterate();
+                if (it.MoveNext() == false)
+                    break;
+                
+                var page = llt.ModifyPage(it.Current);
+                var maybe = new Container(page);
+
+                // we want to ensure that the free list doesnt get too big...
+                // if we don't have space here, we should discard it from the free list
+                // however we need to be sure you are not going to do so when the entries
+                // are abnormally big. In those cases, the reasonable thing to do is just
+                // skip it and create a new page for it but without discarding pages that
+                // would be reasonably used by following requests. 
+                if (!isBigEntry)
```

Comment:
I don't like this.
If we have a big entry, we won't clear the free list and always scan the first 4 pages each time.
A page can drop out and be returned to the free list as data changes on it. Dropping explicitly is a better idea, I think.

Note that our impl means that we _always_ delete and insert data, so no shortage of free list pages.

Reply:
If it is big enough, it will drop all without any recourse. Not a good idea over the long run. Accessing 4 and giving up only if the entry is big enough is a sensible tradeoff to just say, screw it; I will go on creating a new one instead.

Reply:
Really good idea. We clear the free list.
Then we add to that again on additional deletes.

We are *always* doing deletes on updates, so there is no shortage here.

Reply:
Not bought on it, because this is true for Corax, but Container is a general thing; we are using it now for something with a lot of turn-over but these are the kinds of assumptions that we end up tripping on when we find containers users for something else and a client comes with a weird scenario. That's why I would like to be more conservative here.

Reply:
I am not clearing anyone here if we are a big guy. A compromise solution that is acceptable under both scenarios is to limit the amount of tries of big guys, so only a few will get tossed out; before giving up. What do you think?

Reply:
How would you maintain the state?
Another thing to note is that we are handling this on a per single call with all persistent state.
But in our scenario, we are usually involved with a _lot_ of calls here
We can do a heap with the size of the pages we clear, and that should be good enough to not throw too much away

Reply:
@redknightlois note there is question from Oren here.

I've reopened the issue to address this comment: https://issues.hibernatingrhinos.com/issue/RavenDB-18817#focus=Comments-67-370693.0-0

Reply:
A persistent heap would be my go-to solution with no compromise solution involved. Should we open a ticket to handle that post release?

Reply:
A persistent heap is _complex_.
Technically we can use a multi tree, with the key being the size.
Not sure if it is worth it